### ISSUE TITLE
Using racket-open-require-path non-interactively

### ISSUE THREAD

**Issue by btlachance** [greghendershott/racket-mode#161]
Like we discussed on IRC: I'd like to be able to give `racket-open-require-path` (or similar) a single argument rather than having to deal with the minibuffer or the internal functions/transaction queue.

(Maybe this is out of scope for racket-mode, as this isn't necessarily for humans to use in their workflow.)

**Comment by greghendershott** [comment]
What you want to have happen, after you call the function?  Do you want it to return a match with no user interaction at all?  If the answer is yes:

Before I added `racket-open-require-path` in response to a request to emulate how that DrRacket command works, there was a `racket-find-collection` command. This is just a thin wrapper around Asumu Takikawa's `raco-find-collection` package -- which you can install using `raco pkg install raco-find-collection` and use with `raco fc <collection name>`.  Your Elisp code can use this command, too.

**TL;DR** I think the simplest thing for you to do -- which doesn't involve racket-mode, at all -- might be this:

```cl
(defun btlachance/insert-collection-path-at-point (collection)
  "Use raco fc to find the path to COLLECTION and insert it at point."
  (shell-command (format "raco fc %s" collection) t))

;; example use
(btlachance/insert-collection-path-at-point "racket/syntax")
```

If you don't want the result inserted at point, you can grab it from a `*Shell Command Output*` buffer, or some other options: <kbd>C-h f</kbd> enter `shell-command` and check out its docs.

Does that help?

**Comment by btlachance** [comment]
Sorry for not being clear on the result of the function, a list of matches is probably good enough. I only plan to use it to get the path to a single file in a collection, e.g., `typed-racket/infer/infer-unit.rkt`, so maybe something more specific is acceptable.

I'm familiar with raco fc, but unfortunately that doesn't get me quite what I want. If that was able to return a path to a file when given a path like above, then that would be sufficient. Thanks for the tip about the `*Shell Command Output*` buffer, that sounds handy.

In it's simplest form, what I want is: when given a "require path" to a file that exists (according to whatever module open-require-path uses), return the resulting match. I want something that resolves a require path to a path on my filesystem.

Maybe I should just take your suggestion and shell out to `drracket/find-module-path-completions`?

**Comment by greghendershott** [comment]
Ah I see. In that case, the following seems to work fine for me:

```cl
(defun btlachance/find-require-path (path)
  (shell-command
   (format "racket -e \"%S\""
           `(begin
             (require drracket/find-module-path-completions)
             (define f (find-module-path-completions (current-directory)))
             (map car (f ,(format "\"%s\"" path))))))
  (eval (read (get-buffer "*Shell Command Output*"))))

(btlachance/find-require-path "typed/racket")
;; => ("racket" "racket.rkt" "racket")
```

The `%S` formatter is handy, because you can write Racket s-exprs as Elisp sexprs, generally.

**Comment by greghendershott** [comment]
Oops, the `(map car ...)` is wrong, you want the second element of each, with the full pathnames.

```
'((#<path:/usr/racket/share/pkgs/typed-racket-lib/typed/racket>)
  (#<path:/usr/racket/share/pkgs/typed-racket-lib/typed/racket.rkt>)
  (#<path:/usr/racket/share/pkgs/typed-racket-more/typed/racket>))
```
Unfortunately Elisp `read` doesn't know how to read Racket struct literals like `#<path: ...>`. Ah the joys of Racket/Elisp interop.

So really you need to `path->string` all those:

```cl
(defun btlachance/find-require-path (path)
  (shell-command
   (format "/usr/racket/bin/racket -e \"%S\""
           `(begin
             (require drracket/find-module-path-completions)
             (define f (find-module-path-completions (current-directory)))
             (map (lambda (x)
                    (path->string (second x)))
                  (f ,(format "\"%s\"" path))))))
  (eval (read (get-buffer "*Shell Command Output*"))))

(btlachance/find-require-path "typed/racket")
;; =>
;; '("/usr/racket/share/pkgs/typed-racket-lib/typed/racket"
;;   "/usr/racket/share/pkgs/typed-racket-lib/typed/racket.rkt"
;;   "/usr/racket/share/pkgs/typed-racket-more/typed/racket")
```

**Comment by btlachance** [comment]
Thanks so much, Greg!

I was getting close to something like this but yours is so much cleaner. The trick with formatting and s-expressions is extremely handy!
